File: Ballast.Core/src/Client/models/axial-coordinates.ts
import { CubicCoordinates, ICubicCoordinates } from './cubic-coordinates';
import { IOffsetCoordinates, OffsetCoordinates } from './offset-coordinates';

export interface IAxialCoordinates {
    x: number;
    z: number;
}

export class AxialCoordinates implements IAxialCoordinates {

    public readonly x: number;
    public readonly z: number;

    private constructor(state: IAxialCoordinates) {
        this.x = state.x;
        this.z = state.z;
        if (this.x == -0)
            this.x = 0;
        if (this.z == -0)
            this.z = 0;
    }

    public static fromObject(object: IAxialCoordinates) {
        return new AxialCoordinates(object);
    }

    public static fromCubic(object: ICubicCoordinates) {
        return AxialCoordinates.fromObject(object);
    }

    public static fromOffset(object: IOffsetCoordinates) {
        return AxialCoordinates.fromCubic(
            CubicCoordinates.fromOffset(object)
        );
    }

    public static fromOrderedPair(orderedPair: number[]) {
        if (orderedPair.length < 2)
            throw new Error('Length of ordered pair must be 2 (or greater)');
        let x = orderedPair[0];
        let z = orderedPair[1];
        return AxialCoordinates.fromObject({ x: x, z: z });
    }

    public equals(object: IAxialCoordinates) {
        if (!object) {
            return false;
        }
        return (
            this.x == object.x &&
            this.z == object.z
        );
    }

    public equalsCubic(object: ICubicCoordinates) {
        return AxialCoordinates
            .fromCubic(object)
            .equals(this);
    }

    public toCubic() {
        return CubicCoordinates.fromAxial(this);
    }

    public toOffset() {
        return OffsetCoordinates.fromAxial(this);
    }

    public toOrderedPair() {
        return [this.x, this.z];
    }

}

File: Ballast.Core/src/Server/Models/AxialCoordinates.cs
using Ballast.Core.Models.Interfaces;
using System;

namespace Ballast.Core.Models
{
    public class AxialCoordinates : IAxialCoordinates
    {
        public int X { get; private set; }
        public int Z { get; private set; }

        private AxialCoordinates(int x, int z)
        {
            X = x;
            Z = z;
        }

        private AxialCoordinates(IAxialCoordinates state) : this(
            x: state.X, 
            z: state.Z
        ) { }

        public static AxialCoordinates FromProperties(int x, int z) => new AxialCoordinates(
            x: x,
            z: z
        );

        public static AxialCoordinates FromObject(IAxialCoordinates state) =>
            new AxialCoordinates(state);

        public static AxialCoordinates FromCubic(ICubicCoordinates state) =>
            AxialCoordinates.FromObject(state);

        public static AxialCoordinates FromOffset(IOffsetCoordinates state) =>
            AxialCoordinates.FromCubic(
                CubicCoordinates.FromOffset(state)
            );

        public static AxialCoordinates FromOrderedPair(int[] orderedPair)
        {
            if (orderedPair.Length < 2)
                throw new Exception("Length of ordered pair must be 2 (or greater)");
            var x = orderedPair[0];
            var z = orderedPair[1];
            return AxialCoordinates.FromProperties(x: x, z: z);
        }

        public bool Equals(IAxialCoordinates value)
        {
            if (value == null)
                return false;
            return (
                X == value.X &&
                Z == value.Z
            );
        }

        public bool EqualsCubic(ICubicCoordinates state) => 
            AxialCoordinates
                .FromCubic(state)
                .Equals(this);

        public CubicCoordinates ToCubic() =>  CubicCoordinates.FromAxial(this);

        public OffsetCoordinates ToOffset() => OffsetCoordinates.FromAxial(this);

        public int[] ToOrderedPair() => new int[] { X, Z };

    }
}